UnitTests/StreamCopierTest.cs

```diff
@@ -41,191 +40,196 @@ public void SetUp ()
+      var inputMock = new Mock<Stream>();
+      inputMock.Setup (s => s.Read (It.IsAny<Byte[]>(), It.IsAny<int>(), It.IsAny<int>())).Returns (0).Verifiable();
+      inputMock.SetupGet (s => s.CanSeek).Returns (false);
+      inputMock.SetupGet (s => s.Length).Throws (new AssertionException ("Length should not be called, when CanSeek returns false."));
 
-      _streamCopier.CopyStream (inputMock, new MemoryStream());
+      _streamCopier.CopyStream (inputMock.Object, new MemoryStream());
 
-      inputMock.VerifyAllExpectations();
+      inputMock.Verify();
     }
 
     [Test]
     public void CopyStream_PassesCorrectBufferLength ()
     {
-      Stream inputMock = MockRepository.GenerateStrictMock<Stream>();
-      inputMock.Expect (s => s.Read (new byte[_streamCopier.BufferSize], 0, _streamCopier.BufferSize)).Return (0);
-      inputMock.Stub (s => s.CanSeek).Return (false);
+      var inputMock = new Mock<Stream>();
+      inputMock.Setup (s => s.Read (new byte[_streamCopier.BufferSize], 0, _streamCopier.BufferSize)).Returns (0).Verifiable();
+      inputMock.SetupGet (s => s.CanSeek).Returns (false);
 
-      _streamCopier.CopyStream (inputMock, new MemoryStream());
+      _streamCopier.CopyStream (inputMock.Object, new MemoryStream());
 
-      inputMock.VerifyAllExpectations();
+      inputMock.Verify();
     }
 
     [Test]
     public void CopyStream_StreamReturnsMoreDataThanMaximumLength_ThrowsIOException ()
     {
-      Stream inputStub = MockRepository.GenerateStub<Stream>();
-      inputStub.Stub (_ => _.CanSeek).Return (false);
-      inputStub.Expect (_ => _.Read (null, 0, 0)).IgnoreArguments().Return (13);
+      var inputStub = new Mock<Stream>();
+      inputStub.SetupGet (_ => _.CanSeek).Returns (false);
+      inputStub.Setup (_ => _.Read (It.IsAny<Byte[]>(), It.IsAny<int>(), It.IsAny<int>())).Returns (13).Verifiable();
 
       Assert.That (
-          () => _streamCopier.CopyStream (inputStub, new MemoryStream(), 12),
+          () => _streamCopier.CopyStream (inputStub.Object, new MemoryStream(), 12),
           Throws.InstanceOf<IOException>().With.Message.EqualTo ("The stream returned more data (13 bytes) than the specified maximum (12 bytes)."));
     }
 
     [Test]
     public void CopyStream_StreamReturnsMoreDataThanStreamLength_ThrowsIOException ()
     {
-      Stream inputStub = MockRepository.GenerateStub<Stream>();
-      inputStub.Stub (_ => _.CanSeek).Return (true);
-      inputStub.Stub (_ => _.Length).Return (11);
-      inputStub.Stub (_ => _.Read (null, 0, 0)).IgnoreArguments().Return (13).Repeat.Once();
-      inputStub.Stub (_ => _.Read (null, 0, 0)).IgnoreArguments().Return (0);
+      var inputStub = new Mock<Stream>();
+      inputStub.SetupGet (_ => _.CanSeek).Returns (true);
+      inputStub.SetupGet (_ => _.Length).Returns (11);
+      inputStub.SetupSequence (_ => _.Read (It.IsAny<Byte[]>(), It.IsAny<int>(), It.IsAny<int>()))
```

Comment:
That's a nice trick

Reply:
Hmm...considering the other places we use sequence, I do wonder if we should use all APIs available or instead fall back to the explicit sequence API to keep things easier to understand and recognize. Reasoning is, me worrying that we might be using specialized convenicence APIs only one or two times in the code base and then needing to understand one extra API just for this one case. I'm leaning towards changing this to use the more verbose form in hopes of giving future self less mental load to deal with.

Reply:
Personally, I think I would have defaulted to this syntax and only use explicit `MockSequence` objects if the sequence needs to be shared among multiple `Setup` calls with different parameters and mocks. Do you have a gut feeling which case is more common?

Reply:
Regarding what's more common: from what I've gathered by skimming the issue docs, it appears that SetupSequence doesn't work in combination with Verifiable() while the syntax with the MockSequence object does support it? Is this correct? Once that's answered we can further explore this.

Reply:
That is correct, establishing a sequence with a Mocksequence instance still allows calling .Verifiable() at the end, while the SetupSequence syntax does not.

Reply:
I would also prefer the more verbose form (explicit `MockSequence`) as it can be used for sequences of same or different mocks, and still just adds a single line to the test (`var sequence = new MockSequence();`) - the missing ability to call `.Verifiable()` on `.SetupSequence` is a dealbreaker anyway imo.